### PR TITLE
Sync upstream with copy_from_upstram / pqcrystals pull

### PR DESCRIPTION
**Pull request by bhess**
This PR is to show the changes after copy_from_upstream with #880.

It will also be an alternative to #890.

With this configuration, it is pulling the latest Kyber round 3 and Dilithium round 3 from pqcrystals-upstream.

The updates can be reproduced like this:

```
$ ./copy_from_upstream copy
```

To verify that the synced version matches with upstream versions:

```
$ ./copy_from_upstream.py verify
-----
Total schemes: 151 - 151 match with upstream, 0 differ
-----
```

Note that this PR might have to be rebased after #880 is merged.

* [x] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)

Changes to Dilithium Round 3 KAT. Adds Dilithium Round 3 algorithms (2, 3, 5, 2-AES, 3-AES, 5-AES)

* [x] Does this PR change the the list of algorithms available -- either adding or removing?  (If so, PRs in OQS-OpenSSL, OQS-BoringSSL, and OQS-OpenSSH will also be required by the time this is merged.)

### CODE REVIEW

**Review by baentsch — COMMENTED** [review]
I'd somehow prefer automating updates also to these locations: Doable without too much work @bhess ? Maybe @xvzcf has a suggestion for this, too.

**Comment by bhess** [comment]
Thanks for the review. As discussed today, I've added a reminder to the output of #880 to manually update the .md files for now. The docs will be a candidate for further yml'izing and auto-generating later.

**Comment by baentsch** [comment]
> The docs will be a candidate for further yml'izing and auto-generating later.

Agreed. So next steps are then: You resolve conflicts created by the merge of #880 (out of curiosity: why did they happen in the first place?) and update this PR. When CI on the updated/re-based PR pass, I'll do my best to review again/approve tomorrow so we may also have landed this PR same time tomorrow (unless someone else gets around to looking at it quicker). I did create #893 to not forget about the documentation issue.

**Comment by bhess** [comment]
This will be great. I now rebased this branch.

> Agreed. So next steps are then: You resolve conflicts created by the merge of #880 (out of curiosity: why did they happen in the first place?) and update this PR.

I am not sure about this. There seems to be an authentication issue with the address-sanitizer CI. It seems to occur in different independent branches, not only in mine:

https://app.circleci.com/pipelines/github/open-quantum-safe/liboqs/1205/workflows/11205cab-a8a2-4d01-85e3-b5cf91a577f8/jobs/12157
https://app.circleci.com/pipelines/github/open-quantum-safe/liboqs/1204/workflows/cfa94cc5-28f6-4ba3-b58f-4470cb598a14/jobs/12142

**Comment by baentsch** [comment]
> There seems to be an authentication issue with the address-sanitizer CI

Hmm -- according to [this](https://app.circleci.com/pipelines/github/open-quantum-safe/liboqs/1206/workflows/f9c33486-543c-45f7-a560-f045574092d2/jobs/12171/tests) address-sanitizer ran perfectly.

> ps. Reading the comment here, docker might enforce rate limitations: d183ccd

Well, two comments: 1) The reason we commented out the auth sections was the "assumed assurance" that there's no rate limitation for OQS as per this statement:

> No pull rate restrictions will be applied to namespaces approved as non-commercial Open Source projects.

(from https://www.docker.com/blog/what-you-need-to-know-about-upcoming-docker-hub-rate-limiting/)

and 2) The error statement wasn't one listed in https://www.docker.com/increase-rate-limits (but "Error response from daemon: unauthorized: authentication required")

But as an incorrect implementation or wrong classification of our project (or simple alternative facts in press releases :) may be behind this I just re-ran the single failed job via SSH (`alpine-noopenssl`) and everything passed. The weird thing also is that only `alpine-noopenssl` failed while it's basically identical twin `alpine` passed. As I don't have permission to select "Re-Run job" [here](https://app.circleci.com/pipelines/github/open-quantum-safe/liboqs/1206/workflows/f9c33486-543c-45f7-a560-f045574092d2/jobs/12162) I just triggered full CI re-run. If this fails again, we may indeed have to revert https://github.com/open-quantum-safe/liboqs/commit/d183ccda6698a246ddf4e517d6afcd716afbbcf6 . The quickest way to check this would be for you to simply remove the comment marks in front of the auth statements in `.circleci/config.yml`.

**Review by baentsch — COMMENTED** [review]
I didn't find the expected reversion of reference to the bhess repo (instead of main upstreams) e.g., in https://github.com/open-quantum-safe/liboqs/blob/fc35e09a1491fe3ffaba90999c1767e65a694ac9/scripts/copy_from_upstream/copy_from_upstream.yml#L13 but maybe I just got lost in the sea of files :-/ Can you please comment (as to whether this is intentional) before I continue with review?

**Comment by bhess** [comment]
> but maybe I just got lost in the sea of files :-/ Can you please comment (as to whether this is intentional) before I continue with review?

This is correct, it's intentional that it's still pointing to the bhess repo. I'm basically awaiting two PRs to merge it to upstream. The merges basically update the yml-files. I thought that it might make sense to add the option to have liboqs-local versions of the yml-files.
We can as well hold off merging here until the PR is merged.

> The quickest way to check this would be for you to simply remove the comment marks in front of the auth statements in `.circleci/config.yml`

Ok I will do that.

**Comment by baentsch** [comment]
> Yesterday there were a few pipelines running in parallel when it failed.

Yes, I have the gut feeling that was the issue (too many CCI jobs in different PRs running at the same time). Already the re-run I initiated manually earlier also passed: https://github.com/open-quantum-safe/liboqs/pull/891/checks?check_run_id=1828548012

**Comment by baentsch** [comment]
> We can as well hold off merging here until the PR is merged.

That would save us another PR round (going from `bhess` to `pqcrystals`) and in my eyes thus would be preferential -- assuming that doesn't delay things too much.

**Comment by bhess** [comment]
The PRs to pqcrystals have now been merged and copy_from_upstream now points directly to pqcrystals:
https://github.com/open-quantum-safe/liboqs/pull/891/commits/2d6c1ce2c46075646a29473c5dc2818ec3efb449#diff-47bff58854d31ecb37c570e43c7539b0fa5588896d49a457d355655fa8b96f16R13-R23

I echo @jschanck's comment in #890. The currently included AES and fips202 code is indeed redundant. As mentioned in #849 and #890, the handling of common code and imports will have to be further improved. One approach will be to apply patches in copy_from_upstream. Patches could also potentially be avoided by adding shim-api (e.g. for the pqcrystals-namespace) in `src/common`.

Aside from this redundancy, I don't see a functional limitation from this PR right now.

File: src/kem/kyber/kem_kyber_1024.c
Comment (baentsch) [inline src/kem/kyber/kem_kyber_1024.c:47]: This (generated) code seems to contradict the documentation: "**Optimizations**: Portable C with AES, AVX2, POPCNT, SSE2 and SSSE3 instructions (if available at run-time)". Question: Did you build-and-run this branch locally once with OQS_USE_CPU_EXTENSIONS=ON and once with OFF and see the expected performance differences?

File: src/kem/kyber/CMakeLists.txt
Comment (baentsch) [inline src/kem/kyber/CMakeLists.txt:31]: Ahh -- this explains my confusion probably: Does SSSE3 and SSE2 only apply to Kyber90s but not Kyber (-noAES)? If so, let's go ahead and merge.

Comment (bhess) [reply src/kem/kyber/CMakeLists.txt:31]: Yes,  SSSE3 and SSE2 is part of the AES-dependency, so only relevant for Kyber90s.

File: src/kem/kyber/kem_kyber_1024.c
Comment (bhess) [reply src/kem/kyber/kem_kyber_1024.c:47]: Thanks for the suggestion for testing with OQS_USE_CPU_EXTENSIONS=ON/OFF. The performance difference is there as expected.

**Review by baentsch — APPROVED** [review]
Thanks for the additional explanations. I'm fine then.

**Comment by bhess** [comment]
Thanks a lot for the review!

**Pull request merged**
